uqcsbot/api.py

```diff
@@ -305,5 +325,131 @@ def _on_group_joined(self, evt):
     def _on_group_left(self, evt):
         self._on_channel_deleted(evt)
 
-class MembersWrapper(object):
-    pass
+
+class UsersWrapper(object):
+    def __init__(self, bot: 'UQCSBot') -> None:
+        self._bot = bot
+        self._empty_store()
+        self._lock = threading.RLock()
+        self._initialised: bool = False
+
+        self._bind_handlers()
+
+    def _empty_store(self):
+        self._users_by_id: Dict[str, User] = {}
+        self._users_by_display: defaultdict[str, List[User]] = defaultdict(list)
+
+    def _add_user(self, data: dict):
+        user = User.from_dict(data)
+        self._users_by_id[user.user_id] = user
+        self._users_by_display[user.display_name].append(user)
+        return user
+
+    def _initialise(self):
+        with self._lock:
+            if self._initialised:
+                # Fast exit for initialisation while waiting on lock
+                return
+            self._empty_store()
+            for page in self._bot.api.users.list.paginate():
+                for user in page['users']:
+                    self._add_user(user)
+
+    def reload(self):
+        self._initialised = False
+        self._initialise()
+
+    def get(self, display_name_or_id: str, default: T = None, use_cache: bool =True) -> Union['User', List['User'], T]:
+        if use_cache and not self._initialised:
+            self._initialise()
+        if display_name_or_id in self._users_by_id:
+            return self._users_by_id[display_name_or_id]
+        elif display_name_or_id in self._users_by_display:
+            return self._users_by_display[display_name_or_id]
```

Comment:
I just realised, why are we storing users by their display name? like, won't this `get` return a list of users?

Reply:
yup, this will return a list of users (the type reflects that)

Unfortunately display names can collide, but that's there in case we end up needing them for something

Reply:
would prefer to just remove them tbh, unless you have an example where you think its necessary? lets just use user ids